File: swift/problems/206_reverse_linked_list.swift
/**
Given the head of a singly linked list, reverse the list, and return the reversed list.
**/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class Solution {
    func reverseList(_ head: ListNode?) -> ListNode? {
        var prev: ListNode? = nil
        var curr: ListNode? = head
        while curr != nil {
            var nxt: ListNode? = curr?.next
            curr?.next = prev
            prev = curr
            curr = nxt
        }
        return prev
    }
}

File: typescript/problems/206_reverse_linked_list.ts
/**
Given the head of a singly linked list, reverse the list, and return the reversed list.
 */

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function reverseList(head: ListNode | null): ListNode | null {

    let prev: ListNode | null = null

    while (head !== null) {
        let nxt: ListNode | null = head.next;
        head.next = prev
        prev = head
        head = nxt
    }
    return prev
    
};